README.txt
===================
django-downloads
===================

I have a need of access controlled downloads. ie: you must be authenticated
to the site, and given specific download permissions via being a certain
user or in a certain group to be able to download some files. This made
it necessary to do the access control checking inside of django and with the
advent of Django 1.2 and apps like django-guardian we have an interface
that gives us per-object level access controls.

For installation instructions, see the file "INSTALL.txt" in this
directory; for instructions on how to use this application, and on
what it provides, see the file "overview.txt" in the "docs/"
directory.


CHANGELOG.txt
=============================
django-downloads changelog
=============================

Version 0.1, 2010.11.15
--------------------------

* Initial release


downloads/urls.py
#!/usr/bin/env python
#
# File: $Id$
#
"""
URL's for the downloads app.
"""

# system imports
#

# Django imports
#
from django.conf import settings
from django.conf.urls.defaults import *
from django.views.generic.simple import redirect_to
from django.core.urlresolvers import reverse

# Models imports
#
from hackers.downloads.models import Downloadable

# If we have the 'asutils.middleware.RequireLogin' middleware then by
# default all of our views will require users to be logged in. This will
# fail for our rss feeds so, in that case we want to import the wrapper
# functions that we can use to require our rss feeds to use HTTP basic auth
# to authenticate a user. This is a bit backwards as we have to allow
# anonymous wrapped around the function that requires HTTP basic auth.
#
if 'asutils.middleware.RequireLogin' in settings.MIDDLEWARE_CLASSES:
    try:
        from asutils.decorators import view_or_basicauth
        from asutils.middleware import allow_anonymous

        @allow_anonymous
        def basic_auth_feed(request, url, feed_dict = None):
            args = (url,)
            kwargs = { 'feed_dict' : feed_dict }
            return view_or_basicauth(feed, request,
                                     lambda u: u.is_authenticated(),
                                     *args, **kwargs)
    except ImportError:
        def basic_auth_feed(request, url, feed_dict = None):
            return feed(request, url, feed_dict)

urlpatterns = patterns(
    '',

    # url('^$',
    #     redirect_to,
    #     { 'url' : reverse('downloads_list') },
    #     name = "downloads_index"),

    # View all available downloads.
    #
    url('^download/$',
        'hackers.downloads.views.list',
        { 'queryset'    : Downloadable.objects.all(),
          'paginate_by' : 20},
        name = "downloads_list"),

    # Create a new downloadable
    #
    url('^create/$',
        'hackers.downloads.views.create',
        name = 'downloads_create'),

    # View a specific downloadable
    #
    url('^download/(?P<downloadable_id>\d+)/$',
        'hackers.downloads.views.details',
        name = 'downloads_details'),

    # Download a specific downloadable
    #
    url('^download/(?P<downloadable_id>\d+)/download/$',
        'hackers.downloads.views.download',
        name = 'downloads_download'),

    # Edit a downloadable
    #
    url('^download/(?P<downloadable_id>\d+)/edit/$',
        'hackers.downloads.views.edit',
        name = 'downloads_edit'),

    # Delete a downloadable
    #
    url('^download/(?P<downloadable_id>\d+)/delete/$',
        'hackers.downloads.views.delete',
        name = 'downloads_delete'),

    )


downloads/admin.py
#!/usr/bin/env python
#
# File: $Id$
#
"""
Declarations of our models so they appear in the django admin site.
"""

# system imports
#

# Django imports
#
from django.contrib import admin

# Model imports
#
from hackers.downloads.models import Downloadable, Download

##################################################################
##################################################################
#
class DownloadableAdmin(admin.ModelAdmin):
    list_display = ('id', 'blurb', 'content')
    list_filter = ('created', 'mimetype', 'owner')
    search_fields = ('blurb','description')

##################################################################
##################################################################
#
class DownloadAdmin(admin.ModelAdmin):
    list_display = ('id', 'who', 'what', 'when')
    list_filter = ('when', 'who')
    search_fields = ('ip_address','dns_name')

admin.site.register(Downloadable, DownloadableAdmin)
admin.site.register(Download, DownloadAdmin)


downloads/views.py
#!/usr/bin/env python
#
# File: $Id$
#
"""
The views for the download app.
"""

# system imports
#
import datetime
import mimetypes
import os.path
from urllib import quote, unquote

# Django imports
#
from django.http import HttpResponseRedirect, HttpResponseForbidden, Http404
from django.views.generic.list_detail import object_list
from django.shortcuts import get_object_or_404, get_list_or_404
from django.core.urlresolvers import reverse
from django.utils.translation import ugettext_lazy as _
from django.utils.translation import ugettext as _u
from django.db.models import Q
from django import forms

# 3rd party module imports
#
# http://trac.apricot.com/projects/django-asutils
from asutils.utils import asrender_to_response, msg_user
from asutils.sendfile import send_file
from guardian.decorators import permission_required_or_403
from guardian.shortcuts import assign

# Form imports
#
from hackers.downloads.forms import DownloadableForm

# Model imports
#
from hackers.downloads.models import Downloadable, Download

####################################################################
#
def list(request, queryset, paginate_by, extra_context = None,
         template_name = 'downloads/list.html', **kwargs):
    """
    XXX Augment this with search capability like aswiki has

    A list of the downloads the user can see.

    Arguments:
    - `request`: django http request object
    - `queryset`: set of downloads to list
    - `paginate_by`: how many to display on a single page
    - `extra_context`: Dict to add to the request context
    - `template_name`: Template to use for rendering the result.
    """

    # XXX We need to filter out of the object list downloadables
    #     that the user does not have permission to see.
    return object_list(request, queryset, extra_context = extra_context,
                       template_name = template_name, **kwargs)

####################################################################
#
def create(request, extra_context = None, form_class = DownloadableForm,
           template_name = "downloads/create.html", **kwargs):
    """

    Arguments:
    - `request`:
    - `template_name`:
    """

    # Does this user have permission to create downloadloadable objects?
    #
    if not request.user.has_perm("downloads.add_downloadable"):
        return HttpResponseForbidden(_("You do not have permission to create a downloadable item."))

    # On POST we create the uploadable..
    #
    if request.method == 'POST' and 'content' in request.FILES:
        if 'content' in request.FILES:
            file_name = request.FILES['content'].name
            dnldbl = Downloadable(owner = request.user, content = file_name)
            form = form_class(request.POST, request.FILES, instance = dnldbl)
        else:
            dnldbl = Downloadable(owner = request.user, content = file_name)
            form = form_class(request.POST, instance = dnldbl)

        if form.is_valid():
            dnldbl = form.save(commit = False)
            dnldbl.owner = request.user
            dnldbl.save()

            # The owner gets permissions to his own objects.
            #
            assign('downloads.change_downloadable', request.user, dnldbl)
            assign('downloads.delete_downloadable', request.user, dnldbl)
            assign('downloads.view_downloadable', request.user, dnldbl)
            assign('downloads.download_downloadable', request.user, dnldbl)

            msg_user(request.user, _('File "%s" successfully uploaded') % \
                         os.path.basename(dnldbl.content.name))
            return HttpResponseRedirect(dnldbl.get_absolute_url())
    else:
        form = form_class()
    return asrender_to_response(request, template_name, { 'form' : form },
                                extra_context, **kwargs)


####################################################################
#
def details(request, downloadable_id, template_name = 'downloads/details.html',
            extra_context = None, **kwargs):
    """

    Arguments:
    - `request`:
    - `downloadable_id`:
    - `template_name`:
    - `extra_context`:
    - `**kwargs`:
    """
    dnldbl = get_object_or_404(Downloadable, pk = downloadable_id)
    if not request.user.has_perm('downloads.view_downloadable', dnldbl) and \
            request.user != dnldbl.owner:
        return HttpResponseForbidden(_("You do not have permission to view this downloadable item."))

    return asrender_to_response(request, template_name,
                                { 'downloadable' : dnldbl}, extra_context,
                                **kwargs)

####################################################################
#
def download(request, downloadable_id, extra_context = None, **kwargs):
    """

    Arguments:
    - `request`:
    - `downloadable_id`:
    - `extra_context`:
    - `**kwargs`:
    """

    dnldbl = get_object_or_404(Downloadable, pk = downloadable_id)
    if not request.user.has_perm('downloads.download_downloadable', dnldbl) and \
                        request.user != dnldbl.owner:
        return HttpResponseForbidden(_("You do not have permission to "
                                       "download this."))

    # Track that this download was request.
    #
    kwargs = { 'who' : request.user, 'what' : dnldbl }
    if 'REMOTE_ADDR' in request.META:
        kwargs['ip_address'] = request.META['REMOTE_ADDR']
    if 'REMOTE_HOST' in request.META:
        kwargs['dns_name'] = request.META['REMOTE_HOST']

    download = Download(**kwargs)
    download.save()

    # And then send the content to the requester.
    #
    # return send_file(request, dnldbl.content.path,
    #                  content_type = dnldbl.mimetype, blksize = 65536)
    return send_file(request, dnldbl.content.path,
                     content_type = "application/octet-stream", blksize = 65536)

####################################################################
#
def edit(request, downloadable_id, template_name = "downloads/edit.html",
         extra_context = None, form_class = DownloadableForm, **kwargs):
    """

    Arguments:
    - `request`:
    - `downloadable_id`:
    - `template_name`:
    - `extra_context`:
    - `**kwargs`:
    """

    dnldbl = get_object_or_404(Downloadable, pk = downloadable_id)
    if not request.user.has_perm('downloads.change_downloadable', dnldbl) and \
            request.user != dnldbl.owner:
        return HttpResponseForbidden(_("You do not have permission to change this downloadable item."))

    if request.method == "POST":
        # are we also uploading a new file with this edit?
        #
        if 'content' in request.FILES:
            file_name = request.FILES['content'].name
            dnldbl.content = file_name
            form = form_class(request.POST, request.FILES, instance = dnldbl)
        else:
            form = form_class(request.POST, instance = dnldbl)

        if form.is_valid():
            dnldbl = form.save()
            msg_user(request.user, "Downloadable updated.")
            return HttpResponseRedirect(dnldbl.get_absolute_url())
    else:
        form = form_class(instance = dnldbl)

    return asrender_to_response(request, template_name,
                                { 'downloadable' :dnldbl,
                                  'form'    : form },
                                extra_context, **kwargs)

####################################################################
#
def delete(request, downloadable_id, template_name = "downloads/delete.html",
           extra_context = None, **kwargs):
    """

    Arguments:
    - `request`:
    - `downloadable_id`:
    - `template_name`:
    - `extra_context`:
    - `**kwargs`:
    """

    dnldbl = get_object_or_404(Downloadable, pk = downloadable_id)
    if not request.user.has_perm('downloads.delete_downloadable', dnldbl) and \
            request.user != dnldbl.owner:
        return HttpResponseForbidden(_("You do not have permission to delete this downloadable item."))

    if request.method == "POST":
        # Delete the underlying file, and then delete the downloadable.
        #
        name = dnldbl.blurb
        dnldbl.content.delete()
        dnldbl.delete()
        msg_user(request.user, _("Downloadable %s deleted.") % name)
        return HttpResponseRedirect(reverse('downloads_list'))

    return asrender_to_response(request, template_name,
                                { 'downloadable' : dnldbl },
                                extra_context, **kwargs)





downloads/models.py
#
# File: $Id: models.py 1922 2008-11-14 08:26:58Z scanner $
#

# Python standard lib imports
#
import os.path

# Django imports
#
from django.db import models
from django.conf import settings
from django.db.models import permalink
from django.utils.translation import ugettext_lazy as _
from django.core.files.storage import FileSystemStorage
import django.dispatch

# Model imports
#
from django.contrib.auth.models import User
from django.contrib.sites.models import Site

# 3rd party module imports
#
try:
    import tagging
except ImportError:
    tagging = None
try:
    import notification
except ImportError:
    notification = None

# Because downloads serves up downloadable file content and because these bits
# of content may be restricted by permissions we can not have these uploaded
# files sitting under MEDIA_ROOT. If they were someone could retrieve them if
# they knew the path name without going through our permission system at
# all. This is bad, even though serving them up ourselves is expensive I do
# not see where we have a choice as for some people the content may be
# sensitive. We create our own FileSystemStorage object that points at this
# location. Of course, if the setting for DOWNLOADS_UPLOAD_ROOT was not set,
# then we fall back to MEDIA_ROOT.
#
fs = FileSystemStorage(location = getattr(settings, "DOWNLOADS_UPLOAD_ROOT",
                                          os.path.join(settings.MEDIA_ROOT,
                                                       'downloads')))

# We define a signal that is invoked whenever an uploadable is created.
# Primarily used to generate notifications and generate the thumbnail/preview
# image.
#
uploadable_created = django.dispatch.Signal()

MIME_TYPES = (
    ('text/plain', 'plain text'),
    ('application/octet-stream', 'binary file'),
    ('application/zip', 'zip file'),
    ('application/x-gzip', 'gzip file'),
    ('application/x-tar', 'tar file'),
    )

##################################################################
##################################################################
#
class Downloadable(models.Model):
    """
    The core object of this app - a downloadable something or other.
    We treat all of these as simple file attachments.
    We may have a preview image of the attachment.
    """
    owner = models.ForeignKey(User, verbose_name = _('owner'),
                              editable = False, db_index = True,
                              help_text = _('The user that uploaded this '
                                            'file.'))
    blurb = models.CharField(_("blurb"), max_length = 256,
                             help_text = _('A short description of this file.'))
    description = models.TextField(_("description"), max_length = 2048,
                                   blank = True,
                                   help_text = _("A longer description of "
                                                 "this file"))
    content = models.FileField(_('content'), max_length = 1024,
                               null = True,
                               storage = fs, db_index = True,
                               upload_to = "files/%Y/%m/%d")
    created = models.DateTimeField(_('created'), auto_now_add = True,
                                   editable = False, db_index = True,
                                   help_text = _('The time at which '
                                                 'this file was '
                                                 'originally uploaded.'))
    mimetype = models.CharField(_("mime type"), max_length = 128,
                                choices = MIME_TYPES,
                                default = 'application/octet-stream',
                                help_text = _("The mime type of this "
                                              "downloadable file."))
    thumbnail = models.ImageField(_('thumbnail'), max_length = 1024,
                                  storage = fs, null = True, blank = True,
                                  editable = False,
                                  upload_to = "thumbnails/%Y/%m/%d",
                                  help_text = _('A thumbnail preview of '
                                                'the file, if we have a '
                                                'way of rendering it.'))

    class Meta:
        permissions = (
            ('view_downloadable', 'View downloadable'),
            ('download_downloadable', 'Download downloadable'),
            )

    ##################################################################
    #
    def __unicode__(self):
        if self.content:
            return u"<id: %d, File: '%s', mimetype: %s>" % (self.id,
                                                            self.content,
                                                            self.mimetype)
        else:
            return u"<id: %d, File: No file, mimetype: %s>" % (self.id,
                                                               self.mimetype)

    ##################################################################
    #
    @permalink
    def get_absolute_url(self):
        return ('downloads_details', (self.id,))

if tagging:
    # We register the Downloadable model with the tagging module - this gives
    # all of our Downloadables's the ability to have tags.
    #
    tagging.register(Downloadable)

##################################################################
##################################################################
#
class Download(models.Model):
    """
    This class is used to track downloads of a downloadable. Who
    downloaded what, when, how often, and from where.

    We store at least usre, date, ip address and what was downloaded.
    We may also resolve the ip address shortly after the download
    happens and store that separately in the 'hostname' attribute.

    The reason for the hostname as well as the IP address is that
    hostnames change over time and we may care what the host was when
    it happened, not when we need to resolve these names days, weeks
    or months later.
    """
    who = models.ForeignKey(User, verbose_name = _("who"), db_index = True,
                            help_text = _("Who downloaded the file."))
    what = models.ForeignKey(Downloadable, verbose_name = _("what"),
                             db_index = True,
                             help_text = _("What file was downloaded."))
    when = models.DateTimeField(_("when"), auto_now_add = True,
                                editable = False, db_index = True)
    ip_address = models.CharField(_("IP Address"), max_length = 256,
                                  help_text = _("The IP address a file was "
                                                "downloaded from."))
    dns_name = models.CharField(_("DNS Name"), max_length = 1024,
                                null = True,
                                help_text = _("The DNS name of the host "
                                              "a file was downloaded from "
                                              "at the time the download "
                                              "happend"))

    ##################################################################
    #
    def __unicode__(self):
        if self.dns_name and self.dns_name != "":
            return "<Download of '%s' by '%s' at %s from %s (%s)>" % \
                (self.what.download, self.who.username, self.when,
                 self.ip_address, self.dns_name)
        else:
            return "<Download of '%s' by '%s' at %s from %s>" % \
                (self.what.download, self.who.username, self.when,
                 self.ip_address)



downloads/migrations/0002_auto__add_field_downloadable_blurb__add_field_downloadable_description.py
# encoding: utf-8
import datetime
from south.db import db
from south.v2 import SchemaMigration
from django.db import models

class Migration(SchemaMigration):

    def forwards(self, orm):
        
        # Adding field 'Downloadable.blurb'
        db.add_column('downloads_downloadable', 'blurb', self.gf('django.db.models.fields.CharField')(default='', max_length=256), keep_default=False)

        # Adding field 'Downloadable.description'
        db.add_column('downloads_downloadable', 'description', self.gf('django.db.models.fields.TextField')(default='', max_length=2048, blank=True), keep_default=False)

        # Changing field 'Downloadable.content'
        db.alter_column('downloads_downloadable', 'content', self.gf('django.db.models.fields.files.FileField')(max_length=1024, null=True))


    def backwards(self, orm):
        
        # Deleting field 'Downloadable.blurb'
        db.delete_column('downloads_downloadable', 'blurb')

        # Deleting field 'Downloadable.description'
        db.delete_column('downloads_downloadable', 'description')

        # Changing field 'Downloadable.content'
        db.alter_column('downloads_downloadable', 'content', self.gf('django.db.models.fields.files.FileField')(max_length=1024))


    models = {
        'auth.group': {
            'Meta': {'object_name': 'Group'},
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'name': ('django.db.models.fields.CharField', [], {'unique': 'True', 'max_length': '80'}),
            'permissions': ('django.db.models.fields.related.ManyToManyField', [], {'to': "orm['auth.Permission']", 'symmetrical': 'False', 'blank': 'True'})
        },
        'auth.permission': {
            'Meta': {'ordering': "('content_type__app_label', 'content_type__model', 'codename')", 'unique_together': "(('content_type', 'codename'),)", 'object_name': 'Permission'},
            'codename': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            'content_type': ('django.db.models.fields.related.ForeignKey', [], {'to': "orm['contenttypes.ContentType']"}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'name': ('django.db.models.fields.CharField', [], {'max_length': '50'})
        },
        'auth.user': {
            'Meta': {'object_name': 'User'},
            'date_joined': ('django.db.models.fields.DateTimeField', [], {'default': 'datetime.datetime.now'}),
            'email': ('django.db.models.fields.EmailField', [], {'max_length': '75', 'blank': 'True'}),
            'first_name': ('django.db.models.fields.CharField', [], {'max_length': '30', 'blank': 'True'}),
            'groups': ('django.db.models.fields.related.ManyToManyField', [], {'to': "orm['auth.Group']", 'symmetrical': 'False', 'blank': 'True'}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'is_active': ('django.db.models.fields.BooleanField', [], {'default': 'True'}),
            'is_staff': ('django.db.models.fields.BooleanField', [], {'default': 'False'}),
            'is_superuser': ('django.db.models.fields.BooleanField', [], {'default': 'False'}),
            'last_login': ('django.db.models.fields.DateTimeField', [], {'default': 'datetime.datetime.now'}),
            'last_name': ('django.db.models.fields.CharField', [], {'max_length': '30', 'blank': 'True'}),
            'password': ('django.db.models.fields.CharField', [], {'max_length': '128'}),
            'user_permissions': ('django.db.models.fields.related.ManyToManyField', [], {'to': "orm['auth.Permission']", 'symmetrical': 'False', 'blank': 'True'}),
            'username': ('django.db.models.fields.CharField', [], {'unique': 'True', 'max_length': '30'})
        },
        'contenttypes.contenttype': {
            'Meta': {'ordering': "('name',)", 'unique_together': "(('app_label', 'model'),)", 'object_name': 'ContentType', 'db_table': "'django_content_type'"},
            'app_label': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'model': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            'name': ('django.db.models.fields.CharField', [], {'max_length': '100'})
        },
        'downloads.download': {
            'Meta': {'object_name': 'Download'},
            'dns_name': ('django.db.models.fields.CharField', [], {'max_length': '1024', 'null': 'True'}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'ip_address': ('django.db.models.fields.CharField', [], {'max_length': '256'}),
            'what': ('django.db.models.fields.related.ForeignKey', [], {'to': "orm['downloads.Downloadable']"}),
            'when': ('django.db.models.fields.DateTimeField', [], {'auto_now_add': 'True', 'db_index': 'True', 'blank': 'True'}),
            'who': ('django.db.models.fields.related.ForeignKey', [], {'to': "orm['auth.User']"})
        },
        'downloads.downloadable': {
            'Meta': {'object_name': 'Downloadable'},
            'blurb': ('django.db.models.fields.CharField', [], {'max_length': '256'}),
            'content': ('django.db.models.fields.files.FileField', [], {'max_length': '1024', 'null': 'True', 'db_index': 'True'}),
            'created': ('django.db.models.fields.DateTimeField', [], {'auto_now_add': 'True', 'db_index': 'True', 'blank': 'True'}),
            'description': ('django.db.models.fields.TextField', [], {'max_length': '2048', 'blank': 'True'}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'mimetype': ('django.db.models.fields.CharField', [], {'default': "'application/octet-stream'", 'max_length': '128'}),
            'owner': ('django.db.models.fields.related.ForeignKey', [], {'to': "orm['auth.User']"}),
            'thumbnail': ('django.db.models.fields.files.ImageField', [], {'max_length': '1024', 'null': 'True', 'blank': 'True'})
        }
    }

    complete_apps = ['downloads']


downloads/migrations/0001_initial.py
# encoding: utf-8
import datetime
from south.db import db
from south.v2 import SchemaMigration
from django.db import models

class Migration(SchemaMigration):

    def forwards(self, orm):
        
        # Adding model 'Downloadable'
        db.create_table('downloads_downloadable', (
            ('id', self.gf('django.db.models.fields.AutoField')(primary_key=True)),
            ('owner', self.gf('django.db.models.fields.related.ForeignKey')(to=orm['auth.User'])),
            ('content', self.gf('django.db.models.fields.files.FileField')(max_length=1024, db_index=True)),
            ('created', self.gf('django.db.models.fields.DateTimeField')(auto_now_add=True, db_index=True, blank=True)),
            ('mimetype', self.gf('django.db.models.fields.CharField')(max_length=128)),
            ('thumbnail', self.gf('django.db.models.fields.files.ImageField')(max_length=1024, null=True, blank=True)),
        ))
        db.send_create_signal('downloads', ['Downloadable'])

        # Adding model 'Download'
        db.create_table('downloads_download', (
            ('id', self.gf('django.db.models.fields.AutoField')(primary_key=True)),
            ('who', self.gf('django.db.models.fields.related.ForeignKey')(to=orm['auth.User'])),
            ('what', self.gf('django.db.models.fields.related.ForeignKey')(to=orm['downloads.Downloadable'])),
            ('when', self.gf('django.db.models.fields.DateTimeField')(auto_now_add=True, db_index=True, blank=True)),
            ('ip_address', self.gf('django.db.models.fields.CharField')(max_length=256)),
            ('dns_name', self.gf('django.db.models.fields.CharField')(max_length=1024, null=True)),
        ))
        db.send_create_signal('downloads', ['Download'])


    def backwards(self, orm):
        
        # Deleting model 'Downloadable'
        db.delete_table('downloads_downloadable')

        # Deleting model 'Download'
        db.delete_table('downloads_download')


    models = {
        'auth.group': {
            'Meta': {'object_name': 'Group'},
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'name': ('django.db.models.fields.CharField', [], {'unique': 'True', 'max_length': '80'}),
            'permissions': ('django.db.models.fields.related.ManyToManyField', [], {'to': "orm['auth.Permission']", 'symmetrical': 'False', 'blank': 'True'})
        },
        'auth.permission': {
            'Meta': {'ordering': "('content_type__app_label', 'content_type__model', 'codename')", 'unique_together': "(('content_type', 'codename'),)", 'object_name': 'Permission'},
            'codename': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            'content_type': ('django.db.models.fields.related.ForeignKey', [], {'to': "orm['contenttypes.ContentType']"}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'name': ('django.db.models.fields.CharField', [], {'max_length': '50'})
        },
        'auth.user': {
            'Meta': {'object_name': 'User'},
            'date_joined': ('django.db.models.fields.DateTimeField', [], {'default': 'datetime.datetime.now'}),
            'email': ('django.db.models.fields.EmailField', [], {'max_length': '75', 'blank': 'True'}),
            'first_name': ('django.db.models.fields.CharField', [], {'max_length': '30', 'blank': 'True'}),
            'groups': ('django.db.models.fields.related.ManyToManyField', [], {'to': "orm['auth.Group']", 'symmetrical': 'False', 'blank': 'True'}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'is_active': ('django.db.models.fields.BooleanField', [], {'default': 'True'}),
            'is_staff': ('django.db.models.fields.BooleanField', [], {'default': 'False'}),
            'is_superuser': ('django.db.models.fields.BooleanField', [], {'default': 'False'}),
            'last_login': ('django.db.models.fields.DateTimeField', [], {'default': 'datetime.datetime.now'}),
            'last_name': ('django.db.models.fields.CharField', [], {'max_length': '30', 'blank': 'True'}),
            'password': ('django.db.models.fields.CharField', [], {'max_length': '128'}),
            'user_permissions': ('django.db.models.fields.related.ManyToManyField', [], {'to': "orm['auth.Permission']", 'symmetrical': 'False', 'blank': 'True'}),
            'username': ('django.db.models.fields.CharField', [], {'unique': 'True', 'max_length': '30'})
        },
        'contenttypes.contenttype': {
            'Meta': {'ordering': "('name',)", 'unique_together': "(('app_label', 'model'),)", 'object_name': 'ContentType', 'db_table': "'django_content_type'"},
            'app_label': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'model': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            'name': ('django.db.models.fields.CharField', [], {'max_length': '100'})
        },
        'downloads.download': {
            'Meta': {'object_name': 'Download'},
            'dns_name': ('django.db.models.fields.CharField', [], {'max_length': '1024', 'null': 'True'}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'ip_address': ('django.db.models.fields.CharField', [], {'max_length': '256'}),
            'what': ('django.db.models.fields.related.ForeignKey', [], {'to': "orm['downloads.Downloadable']"}),
            'when': ('django.db.models.fields.DateTimeField', [], {'auto_now_add': 'True', 'db_index': 'True', 'blank': 'True'}),
            'who': ('django.db.models.fields.related.ForeignKey', [], {'to': "orm['auth.User']"})
        },
        'downloads.downloadable': {
            'Meta': {'object_name': 'Downloadable'},
            'content': ('django.db.models.fields.files.FileField', [], {'max_length': '1024', 'db_index': 'True'}),
            'created': ('django.db.models.fields.DateTimeField', [], {'auto_now_add': 'True', 'db_index': 'True', 'blank': 'True'}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'mimetype': ('django.db.models.fields.CharField', [], {'max_length': '128'}),
            'owner': ('django.db.models.fields.related.ForeignKey', [], {'to': "orm['auth.User']"}),
            'thumbnail': ('django.db.models.fields.files.ImageField', [], {'max_length': '1024', 'null': 'True', 'blank': 'True'})
        }
    }

    complete_apps = ['downloads']
